presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java

```diff
@@ -0,0 +1,242 @@
+        {
+            parentNodes.push(parentNode);
+        }
+
+        public void popParentNode()
+        {
+            parentNodes.pop();
+        }
+    }
+
+    private static class Rewriter
+            extends SimplePlanRewriter<Context>
+    {
+        private final Session session;
+        private final FunctionAndTypeManager functionAndTypeManager;
+        private final PlanNodeIdAllocator planNodeIdAllocator;
+        private final PlanVariableAllocator planVariableAllocator;
+        private final Map<VariableReferenceExpression, VariableReferenceExpression> keyToRandomKeyMap;
+
+        private Rewriter(Session session,
+                FunctionAndTypeManager functionAndTypeManager, PlanNodeIdAllocator planNodeIdAllocator, PlanVariableAllocator planVariableAllocator,
+                Map<VariableReferenceExpression, VariableReferenceExpression> keyToRandomKeyMap)
+        {
+            this.session = requireNonNull(session, "session is null");
+            this.functionAndTypeManager = requireNonNull(functionAndTypeManager, "functionAndTypeManager is null");
+            this.planNodeIdAllocator = requireNonNull(planNodeIdAllocator, "planNodeIdAllocator is null");
+            this.planVariableAllocator = requireNonNull(planVariableAllocator, "planVariableAllocator is null");
+            this.keyToRandomKeyMap = requireNonNull(keyToRandomKeyMap, "keyToRandomKeyMap is null");
+        }
+
+        private static boolean isSupportedType(VariableReferenceExpression variable)
+        {
+            return Stream.of(VARCHAR, BIGINT, DATE).anyMatch(x -> x.equals(variable.getType()));
+        }
+
+        private RowExpression randomizeJoinKey(RowExpression keyExpression, String prefix)
+        {
+            int partitionCount = getHashPartitionCount(session);
+            RowExpression randomNumber = call(
+                    functionAndTypeManager,
+                    "random",
+                    BIGINT,
+                    constant((long) partitionCount, BIGINT));
+            RowExpression randomNumberVarchar = call("CAST", functionAndTypeManager.lookupCast(CAST, randomNumber.getType(), VARCHAR), VARCHAR, randomNumber);
+            RowExpression concatExpression = call(functionAndTypeManager,
+                    "concat",
+                    VARCHAR,
+                    ImmutableList.of(constant(Slices.utf8Slice(prefix), VARCHAR), randomNumberVarchar));
+
+            RowExpression castToVarchar = call("CAST", functionAndTypeManager.lookupCast(CAST, keyExpression.getType(), VARCHAR), VARCHAR, keyExpression);
+            return new SpecialFormExpression(SpecialFormExpression.Form.COALESCE, VARCHAR, ImmutableList.of(castToVarchar, concatExpression));
+        }
+
+        // Do not need to generate randomized variable if the joinKey 1) has already been randomized 2) included in the output of the source node
+        private boolean randomKeyAvailable(PlanNode source, VariableReferenceExpression joinKey)
```

Comment:
Name it something like: isAlreadyRandomized

Reply:
Done